extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -278,13 +277,18 @@ private boolean shouldAutoRedirect(TenantConfigContext configContext, RoutingCon
                 // Local redirect restoring the original request path, the state cookie is no longer needed
                 removeCookie(context, configContext, getStateCookieName(configContext));
             }
+        } else if (configContext.oidcConfig.authentication.isChallengeWithoutStateCookie()) {
+            LOG.debug("The state cookie is missing after a redirect from IDP, re-authentication is required");
+            return Uni.createFrom().optional(Optional.empty());
         } else {
             // State cookie must be available to minimize the risk of CSRF
-            LOG.debug("The state cookie is missing after a redirect from IDP");
+            LOG.debug("The state cookie is missing after a redirect from IDP, authentication has failed");
```

Comment:
Isn't just a matter of checking whether or not the state cookie was set and if not just redirect to the IdP ?

Reply:
@pedroigor The current behaviour in Quarkus is what is happening in this branch in this PR, if the state cookie is absent when Quarkus OIDC thinks it is processing a code response from the IDP redirect then it must be 401. Without it we were getting a lot of reports of the browser going into the redirection loop :-) due to the cookie being lost after the redirect, the user is redirected back and the IDP redirects the user back again, etc, so it is a safety check which should remain a default option IMHO.
May be you meant a session cookie ? It is still the case that if it is not available then the user is redirected to IDP.

Reply:
As you said, if the state cookie is absent it may think it is processing a code. But if the state cookie is not set, that means the code flow even started so we could just force the flow to start.

Maybe I should try this out and see if I'm missing something.

Reply:
Hi Pedro @pedroigor `if the state cookie is absent it may think it is processing a code. But if the state cookie is not set...` it is the same situation :-).
So, putting this PR aside for the moment, when we have a `code` parameter set we assume it is part of the code flow response and therefore we expect that the `state` cookie is present. If it is not present, then until this PR, it is 401 (risk of the redirect loops otheriwise if we return the user back to IDP again - typically it is happening when say you access `index.html` on the 1st request and then `/web-app/service` on the redirect from IDP, etc, and if no cookie path/domain property is set).
The only cookie which we have always used to decide whether to redirect to IDP or not, based on its presence/absence is the session cookie. We only use the state cookie to coordinate the code flow.

And with this PR, we have a situation where a code is coming on the 1st unauthenticated request, or, as it appears in the user case, after the session has expired, so yes, with the new option, we will give the users an option to redirect the users to IDP if the state cookie is absent to resolve the ambiguity with the custom `code` parameter vs a `code` returned by IDP.

Does it clarify it :-) ?